tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,263 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
   return memoized_lookup;
 }
 
+bool IsConstantExpression(const HloInstruction* op,
+                          absl::flat_hash_map<const HloInstruction*, bool> &record) {
+  if (record.find(op) != record.end()) {
+    return record[op];
+  }
+  if (Match(op, m::Constant())) {
+    record[op] = true;
+    return true;
+  }
+  if (Match(op, m::Parameter())) {
+    record[op] = false;
+    return false;
+  }
+  for (auto operand : op->operands()) {
+    if (!IsConstantExpression(operand, record)) {
+      record[op] = false;
+      return false;
+    }
+  }
+  record[op] = true;
+  return true;
+}
+
+StatusOr<HloInstruction*>
+AlgebraicSimplifierVisitor::OptimizeDotOfReorderContractingDims(
+    HloInstruction* dot) {
+  HloInstruction* reshape = nullptr;
+  HloInstruction* transpose = nullptr;
+  HloInstruction* input = nullptr;
+  HloInstruction* constant_expr = nullptr;
+  if (!Match(dot, m::Dot().WithBinaryOperandsAnyOrder(
+                      m::Reshape(&reshape,
+                                 m::Transpose(&transpose,
+                                              m::Op(&input).IsNonConstant())),
```

Comment:
I will remove the IsNonConstant pattern. This is again due to my incorrect understanding of the optimization pass. I thought we only iterative on algsimp pass, then allowing constant on both side would cause a ping-pong situation.